File: go/aontumodel_test.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"
)

// Every .aon is a module, named by its path; a file named after the
// directory holding it collapses, so aontu/code/code.aon is
// `aontu:code` and aontu/render/lang/go.aon is `aontu:render/lang/go`.
func walkModels(t *testing.T, dir, rel string, found []string) []string {
	entries, err := os.ReadDir(dir)
	if nil != err {
		t.Fatal(err)
	}
	for _, entry := range entries {
		subrel := entry.Name()
		if "" != rel {
			subrel = rel + "/" + entry.Name()
		}
		if entry.IsDir() {
			found = walkModels(t, filepath.Join(dir, entry.Name()), subrel, found)
			continue
		}
		if !strings.HasSuffix(entry.Name(), ".aon") {
			continue
		}
		part := strings.Split(strings.TrimSuffix(subrel, ".aon"), "/")
		if 1 < len(part) && part[len(part)-1] == part[len(part)-2] {
			part = part[:len(part)-1]
		}
		found = append(found, aontuScheme+strings.Join(part, "/"))
	}
	return found
}

// go/aontumodel/ is a committed mirror because //go:embed cannot read
// above its own package directory, and a copy nothing compares is a second
// source of truth waiting to drift.
func TestAontuModelsAreTheCanonicalTree(t *testing.T) {
	if 0 == len(aontuModels) {
		t.Fatal("no models: has the generator run?")
	}
	norm := func(s string) string { return strings.ReplaceAll(s, "\r\n", "\n") }
	for _, name := range aontuModels {
		rel := strings.Split(strings.TrimPrefix(name, aontuScheme), "/")
		deep := append([]string{"..", "aontu"}, rel...)
		deep = append(deep, rel[len(rel)-1]+".aon")
		at := filepath.Join(deep...)
		if _, err := os.Stat(at); nil != err {
			at = filepath.Join(append([]string{"..", "aontu"}, rel...)...) + ".aon"
		}
		shared, err := os.ReadFile(at)
		if nil != err {
			t.Fatal(err)
		}
		if norm(string(shared)) != norm(aontuSources[name]) {
			t.Fatalf("%s is stale against aontu/ - run `make aontu`", name)
		}
	}
}

// A model added to the tree and not regenerated is drift the table
// alone cannot see.
func TestAontuTreeServesExactlyTheListedModels(t *testing.T) {
	found := walkModels(t, filepath.Join("..", "aontu"), "", nil)
	sort.Strings(found)
	if strings.Join(found, ",") != strings.Join(aontuModels, ",") {
		t.Fatalf("aontu/ and aontuModels disagree - run `make aontu`: %v vs %v",
			found, aontuModels)
	}
	if !sort.StringsAreSorted(aontuModels) {
		t.Fatalf("aontuModels is not sorted: %v", aontuModels)
	}
	if len(aontuSources) != len(aontuModels) {
		t.Fatalf("aontuSources has %d entries for %d models",
			len(aontuSources), len(aontuModels))
	}
	for _, name := range aontuModels {
		if !strings.HasPrefix(name, aontuScheme) {
			t.Fatalf("%s does not carry the scheme", name)
		}
		if "" == aontuSources[name] {
			t.Fatalf("%s has no source", name)
		}
	}
}

File: ts/test/aontumodel.test.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


import { describe, test } from 'node:test'
import * as Assert from 'node:assert'
import * as Fs from 'node:fs'
import * as Path from 'node:path'

import { AONTU_SCHEME, AONTU_SOURCES, AONTU_MODELS } from '../dist/aontumodel'


const REPO = Path.join(__dirname, '..', '..')
const TREE = Path.join(REPO, 'aontu')


// A module is a directory holding a file named after it: aontu/lang/go
// holds go.aon and is `aontu:lang/go`.
function walk(dir: string, rel: string, found: string[]): string[] {
  for (const entry of Fs.readdirSync(dir, { withFileTypes: true })) {
    if (!entry.isDirectory()) { continue }
    const subrel = rel ? rel + '/' + entry.name : entry.name
    if (Fs.existsSync(Path.join(dir, entry.name, entry.name + '.aon'))) {
      found.push(AONTU_SCHEME + subrel)
    }
    walk(Path.join(dir, entry.name), subrel, found)
  }
  return found
}


function leafOf(name: string): string {
  const rel = name.slice(AONTU_SCHEME.length).split('/')
  return Path.join(TREE, ...rel, rel[rel.length - 1] + '.aon')
}


describe('aontumodel', () => {

  // A generated copy that nothing compares is a second source of truth
  // waiting to drift.
  test('sources-are-identical-with-the-tree', () => {
    Assert.ok(0 < AONTU_MODELS.length, 'no models: has the generator run?')
    for (const name of AONTU_MODELS) {
      const text = Fs.readFileSync(leafOf(name), 'utf8')
        .replaceAll('\r\n', '\n').replaceAll('\r', '\n')
      Assert.equal(AONTU_SOURCES[name], text,
        name + ' is stale against aontu/ — run `make aontu`')
    }
  })

  test('tree-serves-exactly-the-listed-models', () => {
    Assert.deepEqual(AONTU_MODELS, walk(TREE, '', []).sort(),
      'aontu/ and AONTU_MODELS disagree — run `make aontu`')
    Assert.deepEqual(AONTU_MODELS, [...AONTU_MODELS].sort(),
      'AONTU_MODELS is not sorted')
    Assert.deepEqual(Object.keys(AONTU_SOURCES).sort(), AONTU_MODELS)
  })

  test('every-name-carries-the-scheme', () => {
    for (const name of AONTU_MODELS) {
      Assert.ok(name.startsWith(AONTU_SCHEME), name)
    }
  })
})
